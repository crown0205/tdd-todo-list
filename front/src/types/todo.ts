interface Todo {
  id: number;
  title: string;
  isCompleted: boolean;
}

export type { Todo };
