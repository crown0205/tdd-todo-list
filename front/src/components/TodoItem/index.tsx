import { Todo } from '../../types/todo';

interface TodoItemProps {
  todo: Todo;
  onToggle: (id: number) => void;
  onDelete: (id: number) => void;
}

export default function TodoItem({ todo, onToggle, onDelete }: TodoItemProps) {
  return (
    <div
      className="flex items-center gap-3 p-3 bg-gray-50 rounded-lg border border-gray-200"
      data-testid="todo-item"
    >
      <input
        type="checkbox"
        className="w-5 h-5"
        checked={todo.isCompleted}
        onChange={() => onToggle(todo.id)}
        data-testid="todo-checkbox"
      />
      <span
        className={`flex-1 ${
          todo.isCompleted ? 'line-through text-gray-500' : ''
        }`}
      >
        {todo.title}
      </span>

      <button
        className="text-red-500 hover:text-red-600"
        data-testid="delete-button"
        onClick={() => onDelete(todo.id)}
      >
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className="h-5 w-5"
          viewBox="0 0 20 20"
          fill="currentColor"
        >
          <path
            fillRule="evenodd"
            d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
            clipRule="evenodd"
          />
        </svg>
      </button>
    </div>
  );
}
