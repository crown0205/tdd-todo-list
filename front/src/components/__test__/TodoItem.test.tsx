import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import { Todo } from '../../types/todo';
import TodoItem from '../TodoItem/index';

describe('TodoItem', () => {
  const mockTodo: Todo = {
    id: 1,
    title: '테스트 할일',
    isCompleted: false,
  };

  const mockOnToggle = vi.fn();
  const mockOnDelete = vi.fn();

  beforeEach(() => {
    mockOnToggle.mockClear();
    mockOnDelete.mockClear();
  });

  test('할일 항목이 정상적으로 렌더링되어야 합니다', () => {
    render(
      <TodoItem
        todo={mockTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );

    expect(screen.getByTestId('todo-item')).toBeInTheDocument();
    expect(screen.getByTestId('todo-title')).toHaveTextContent('테스트 할일');
    expect(screen.getByTestId('todo-checkbox')).not.toBeChecked();
  });

  test('TodoItem 스냅샷이 일치해야 합니다', () => {
    const { container } = render(
      <TodoItem
        todo={mockTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  test('완료된 할일은 체크박스가 체크되어 있어야 합니다', () => {
    const completedTodo = { ...mockTodo, isCompleted: true };

    render(
      <TodoItem
        todo={completedTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );

    expect(screen.getByTestId('todo-checkbox')).toBeChecked();
    expect(screen.getByTestId('todo-title')).toHaveClass('line-through');
  });

  test('체크박스 클릭 시 onToggle이 호출되어야 합니다', () => {
    render(
      <TodoItem
        todo={mockTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );

    fireEvent.click(screen.getByTestId('todo-checkbox'));
    expect(mockOnToggle).toHaveBeenCalledWith(mockTodo.id);
  });

  test('삭제 버튼 클릭 시 onDelete가 호출되어야 합니다', () => {
    render(
      <TodoItem
        todo={mockTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );

    fireEvent.click(screen.getByTestId('delete-button'));
    expect(mockOnDelete).toHaveBeenCalledWith(mockTodo.id);
  });

  test('키보드로 할일을 완료/삭제할 수 있어야 합니다', () => {
    render(
      <TodoItem
        todo={mockTodo}
        onToggle={mockOnToggle}
        onDelete={mockOnDelete}
      />,
    );

    const checkbox = screen.getByTestId('todo-checkbox');
    fireEvent.keyDown(checkbox, { key: 'Enter', code: 'Enter' });
    expect(mockOnToggle).toHaveBeenCalled();

    const deleteButton = screen.getByTestId('delete-button');
    fireEvent.keyDown(deleteButton, { key: 'Enter', code: 'Enter' });
    expect(mockOnDelete).toHaveBeenCalled();
  });
});
