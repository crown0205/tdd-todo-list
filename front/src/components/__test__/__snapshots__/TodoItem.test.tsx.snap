// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem > TodoItem 스냅샷이 일치해야 합니다 1`] = `
<div>
  <div
    class="flex items-center gap-3 p-3 bg-gray-50 rounded-lg border border-gray-200"
    data-testid="todo-item"
  >
    <input
      class="w-5 h-5"
      data-testid="todo-checkbox"
      type="checkbox"
    />
    <span
      class="flex-1 "
      data-testid="todo-title"
    >
      테스트 할일
    </span>
    <button
      class="text-red-500 hover:text-red-600"
      data-testid="delete-button"
    >
      <svg
        class="h-5 w-5"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
